### PR TITLE
Cancel Payments

### PR DESCRIPTION
**Pull request by acreilly**
https://github.com/boomerdigital/solidus_amazon_payments/issues/26

### CODE REVIEW

**Comment by jordan-brough** [comment]
@acreilly thanks for tackling this!

It would be great to avoid monkey-patching `Spree::Payment::Processing` if possible.  I wonder if we ought to be setting the response code on these payments, and if we did if that might provide a clean way to get at the data that we need?

File: app/models/spree/payment/processing_decorator.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+Spree::Payment::Processing.module_eval do
+  def cancel!
```
Comment (manmartinez) [inline app/models/spree/payment/processing_decorator.rb]: Yeah like Jordan suggests it would be great if we could avoid this, one idea that comes to mind is using the `response_code`, I believe the response code is recorded on the payment, so we should be able to fetch the payment using it, from there we can get the amazon_transaction using `payment.source` and the order using `payment.order` :)

File: app/models/spree/payment_decorator.rb [hunk 2]
```diff
@@ -9,4 +9,9 @@
 ##
 Spree::Payment.class_eval do
   scope :amazon, ->{ where(source_type: 'Spree::AmazonTransaction') }
+
+
+  def amazon_transaction?
```
Comment (manmartinez) [inline app/models/spree/payment_decorator.rb]: Do we still need this method?

Comment (acreilly) [inline app/models/spree/payment_decorator.rb]: Nope, removed it and added spec for the scope instead

File: app/models/spree/gateway/amazon.rb [hunk 3]
```diff
@@ -145,6 +145,16 @@ def void(response_code, gateway_options)
       return ActiveMerchant::Billing::Response.new(true, "Success", response)
     end
 
+    def cancel(response_code)
+      payment = Spree::Payment.find_by(response_code: response_code)
+      order = payment.order
+      load_amazon_mws(payment.source.order_reference)
+
+      response = @mws.refund(response_code, order.number, order.total, order.currency)
```
Comment (jordan-brough) [inline app/models/spree/gateway/amazon.rb]: FYI I don't think `order.total` will always be the correct thing to use here.  But since that's what `#void` is already using I think that would be fine to leave for another PR.  I've opened an issue for that here: https://github.com/boomerdigital/solidus_amazon_payments/issues/34

Comment (jordan-brough) [inline app/models/spree/gateway/amazon.rb]: Could we possibly use `"#{order.number}-cancel` for the second parameter?  Just as a nicety to make things more understandable when looking in the Amazon admin & etc.

File: app/models/spree/amazon_transaction.rb [hunk 4]
```diff
@@ -10,6 +10,7 @@
 module Spree
   class AmazonTransaction < ActiveRecord::Base
     has_many :payments, :as => :source
+    belongs_to :order
```
Comment (jordan-brough) [inline app/models/spree/amazon_transaction.rb]: We can get rid of this now right?

File: app/models/spree/gateway/amazon.rb [hunk 5]
```diff
@@ -105,7 +105,7 @@ def capture(amount, amazon_checkout, gateway_options={})
       t = order.amazon_transaction
       t.capture_id = response.fetch("CaptureResponse", {}).fetch("CaptureResult", {}).fetch("CaptureDetails", {}).fetch("AmazonCaptureId", nil)
       t.save!
-      return ActiveMerchant::Billing::Response.new(response.fetch("CaptureResponse", {}).fetch("CaptureResult", {}).fetch("CaptureDetails", {}).fetch("CaptureStatus", {})["State"] == "Completed", "OK", response)
+      return ActiveMerchant::Billing::Response.new(response.fetch("CaptureResponse", {}).fetch("CaptureResult", {}).fetch("CaptureDetails", {}).fetch("CaptureStatus", {})["State"] == "Completed", "OK", response, {authorization: t.capture_id})
```
Comment (jordan-brough) [inline app/models/spree/gateway/amazon.rb:146]: This is great. Could we add a spec somewhere that verifies that `spree_payments.response_code` does get set after performing a capture?

**Comment by acreilly** [comment]
@jordan-brough @manmartinez Please review.

File: app/models/spree/gateway/amazon.rb [hunk 6]
```diff
@@ -183,6 +183,16 @@ def void(response_code, gateway_options)
       return ActiveMerchant::Billing::Response.new(true, "Success", response)
     end
 
+    def cancel(response_code)
+      payment = Spree::Payment.find_by(response_code: response_code)
```
Comment (jordan-brough) [inline app/models/spree/gateway/amazon.rb]: Nitpick: Let's use `find_by!` instead of `find_by` so that we fail with a clearer error message if we can't find it?

File: app/models/spree/gateway/amazon.rb [hunk 7]
```diff
@@ -183,6 +183,16 @@ def void(response_code, gateway_options)
...
+      order = payment.order
+      load_amazon_mws(payment.source.order_reference)
+
+      response = @mws.refund(response_code, order.number, payment.credit_allowed, payment.currency)
```
Comment (jordan-brough) [inline app/models/spree/gateway/amazon.rb]: I think we actually want to do the same thing here as we're doing in `void` -- either cancel or refund.  That's what [solidusio/solidus_braintree](https://github.com/solidusio/solidus_braintree/blob/81176dcdab70390870e2111be8ec1933b99782a9/app/models/solidus/gateway/braintree_gateway.rb#L169-L177) does at least.  It might actually be that we're doing an odd thing with `void` (perhaps `void` should only cancel and not refund).

But in any case for now maybe we could extract that common code for both of these methods and do the same thing in both places?

**Review by jordan-brough — CHANGES_REQUESTED** [review]
@acreilly this is looking great, thank you again for tackling this! I left just a couple comments.

File: app/models/spree/gateway/amazon.rb (hunk 7 above)
Comment (acreilly) [reply app/models/spree/gateway/amazon.rb]: But all we are given is response code, how do we know if we have to cancel or refund? The only time cancel is called is when an admin cancels an order. The cancel! method is only called on payments that have been captured, so to refund would be the only option available.

https://github.com/solidusio/solidus/blob/master/core/app/models/spree/order.rb#L762-L769

Comment (jordan-brough) [reply app/models/spree/gateway/amazon.rb]: > all we are given is response code, how do we know if we have to cancel or refund?

We can do the same thing that `def void` is doing, just above this `def cancel` right?

> The cancel! method is only called on payments that have been captured

Great catch about [Order#after_cancel](https://github.com/solidusio/solidus/blob/ebda5c8/core/app/models/spree/order.rb#L762-L769).  I think that might be a bug in Solidus.  I've opened a PR to find out: https://github.com/solidusio/solidus/pull/1529

Either way, `Payment#cancel` is an interface that is callable outside of `Order#after_cancel`.  Shouldn't calling `cancel` on an authorized Amazon payment cause it to be voided?  Other Solidus code I can find does that.  E.g. [the Braintree example](https://github.com/solidusio/solidus_braintree/blob/81176dc/app/models/solidus/gateway/braintree_gateway.rb#L169-L177) I mentioned and also [StoreCredit in Solidus](https://github.com/solidusio/solidus/blob/v2.0.0/core/app/models/spree/payment_method/store_credit.rb#L81-L85).

**Comment by jordan-brough** [comment]
FYI: Github is hiding it as "outdated" but I've left another comment [here](https://github.com/boomerdigital/solidus_amazon_payments/pull/30#discussion_r83785289).

File: app/models/spree/gateway/amazon.rb (hunk 7 above)
Comment (acreilly) [reply app/models/spree/gateway/amazon.rb]: I see where you are coming from. Here is the commit: https://github.com/boomerdigital/solidus_amazon_payments/pull/30/commits/af7b0688b4bc86147dcf862538cad32f7c60f65b

Comment (jordan-brough) [reply app/models/spree/gateway/amazon.rb]: @acreilly thanks!

File: app/models/spree/gateway/amazon.rb [hunk 8]
```diff
@@ -183,6 +183,21 @@ def void(response_code, gateway_options)
       return ActiveMerchant::Billing::Response.new(true, "Success", response)
     end
 
+    def cancel(response_code)
+      payment = Spree::Payment.find_by!(response_code: response_code)
+      order = payment.order
+      load_amazon_mws(payment.source.order_reference)
```
Comment (jordan-brough) [inline app/models/spree/gateway/amazon.rb:189]: It feels like once we get the `order`, both `void` and `cancel` could use the exact same code, couldn't they? i.e. could we extract everything after `order =` into a method shared by both `void` and `cancel`?

File: app/models/spree/gateway/amazon.rb [hunk 9]
```diff
@@ -183,6 +183,21 @@ def void(response_code, gateway_options)
...
+      capture_id = order.amazon_transaction.capture_id
+
+      if capture_id.nil?
+        response = @mws.cancel(order.amazon_transaction.order_reference)
+      else
+        response = @mws.refund(response_code, order.number, payment.credit_allowed, payment.currency)
+      end
+
+      return ActiveMerchant::Billing::Response.new(true, "#{order.number}-cancel", response)
```
Comment (jordan-brough) [inline app/models/spree/gateway/amazon.rb:198]: FYI regarding the `"#{order.number}-cancel"` here:  I had requested something like that [here](https://github.com/boomerdigital/solidus_amazon_payments/pull/30#r75179056), but I was referencing the call to `@mws.refund`, not for the `Response.new`.  I don't think the second parameter to `Response.new` is really used for anything in Solidus.

**Review by jordan-brough — APPROVED** [review]
Looks good to me! I left a few minor comments but am 👍 on the PR whether those get changed or not.
